cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    free(command);
+    
+    if (io.write_fd == 0 || io.read_fd == 0)
+    {
+        Log(LOG_LEVEL_VERBOSE, "Error occurred while opening pipes for "
+            "communication with package module.");
+        return -1;
+    }
+    
+    int res = 0;
+    if (WriteScriptData(data, &io) != strlen(data))
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Was not able to send whole data to package module.");
+        res = -1;
+    }
+    
+    /* If script returns non 0 status */
+    if (cf_pclose_full_duplex(&io) != EXIT_SUCCESS)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Package module returned with non zero return code.");
+        res = -1;
+    }
+    return res;
+}
+
+/* In some cases the response is expected to be not filled out. Some requests
+   will have response filled only in case of errors. */
+static int ReadWriteDataToPackageScript(const char *args, const char *request,
+                                        Rlist **response,
+                                        const PackageModuleWrapper *wrapper)
+{
+    assert(args && wrapper);
+    
+    char *command = StringFormat("%s %s", wrapper->path, args);
+    IOData io = cf_popen_full_duplex(command, false);
+    free(command);
+    
+    if (io.write_fd == 0 || io.read_fd == 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating "
+            "package module.");
+        return -1;
+    }
+    
+    if (WriteScriptData(request, &io) != strlen(request))
+    {
+        Log(LOG_LEVEL_VERBOSE, "Couldn't send whole data to package module.");
+        return -1;
+    }
+    
+    /* We can have some error message here. */
+    Rlist *res = ReadDataFromPackageScript(&io);
+    
+    /* If script returns non 0 status */
+    if (cf_pclose_full_duplex(&io) != EXIT_SUCCESS)
+    {
+        Log(LOG_LEVEL_VERBOSE,
+            "Package module returned with non zero return code.");
+        RlistDestroy(res);
+        return -1;
+    }
+    
+    *response = res;
+    return 0;
+}
+
+static void FreePackageInfo(PackageInfo *package_info)
+{
+    if (package_info)
+    {
+        free(package_info->arch);
+        free(package_info->name);
+        free(package_info->version);
+
+        free(package_info);
+    }
+}
+
+static char *ParseOptions(Rlist *options)
+{
+    if (RlistIsNullList(options))
+    {
+        return SafeStringDuplicate("");
+    }
+    
+    Buffer *data = BufferNew();
+    for (Rlist *rp = options; rp != NULL; rp = rp->next)
+    {
+        char *value = RlistScalarValue(rp);
+        BufferAppendString(data, "options=");
+        BufferAppendString(data, value);
+        BufferAppendString(data, "\n");
+    }
+    return BufferClose(data);
+}
+
+static PackageInfo *ParseAndCheckPackageDataReply(const Rlist *data)
+{
+    PackageInfo * package_data = xcalloc(1, sizeof(PackageInfo));
+    
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
```

Comment:
There would be a certain virtue to using `const char *` here and for `type` below.
You don't intend to actually modify it; and declaring it `const` makes it immediately obvious that we don't have ownership of (hence a duty to free) it.